design/one-pager-error-and-event-reporting.md

```diff
@@ -0,0 +1,73 @@
+# Error and event reporting
+
+* Owner: Daniel Suskin (@suskin)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Introduction
+
+*Why does this document exist?*
+
+The question of debuggability and troubleshooting has started to come up
+recently, as we have tried out some scenarios that include Stacks and
+different providers. The intent of this document is to propose a mindset
+and framework for reporting errors and other information.
+
+## Goals
+
+In short, a non-admin user and an admin user should both be able to
+debug any issues only by inspecting logs and events. There should be no
+need to rebuild the Crossplane binary or to reach out to a Crossplane
+developer.
+
+A user should be able to:
+
+* Debug an issue without rebuilding the Crossplane binary
+* Understand an issue without contacting a cluster admin
+* Ask a cluster admin to check the logs for more details about the
+  reason the issue happened, if the details are not part of the error
+  message
+
+A cluster admin should be able to:
+
+* Debug an issue without rebuilding the Crossplane binary
+* Debug an issue only by looking at the logs
+* Debug an issue without needing to contact a Crossplane developer
+
+
+## Error reporting in the logs
+
+Error reporting in the logs is mostly intended for consumption by
+cluster admins. A cluster admin should be able to debug any issue by
+inspecting the logs, without needing to add more logs themselves or
+contact a Crossplane developer. This means that logs should contain:
+
+* Error messages
+* Any context leading up to an error, so that the errors can be debugged
+  (the context can be logged at a more verbose level)
+* Ideally, a unique identifier per session so that log lines can be
+  tracked as a sequence when there are multiple workers logging to the
+  same file
```

Comment:
I agree to some extent, but not entirely. See the long comment about scenarios for more context, but I believe I was imagining the unique identifier as a way to aggregate the relevant information for an error event just by looking at the logs, and without needing to depend on some other system to do the aggregation. However, I suppose one could also think of it as tracing a reconciliation.

Reply:
> However, I suppose one could also think of it as tracing a reconciliation.

Yeah, that was my thought here. Adding a unique ID to log lines is handy and lightweight compared to a full blown tracing system, but we should be careful to avoid crossing over into building a tracing system if we go down this path.

There has been some talk of switching the controller-runtime from Prometheus to Opencensus telemetry in https://github.com/kubernetes-sigs/controller-runtime/issues/305. Opencensus is designed (very nicely IMO) such that instrumenting for timeseries and tracing has a lot of overlap. This could be a cheap (engineering effort wise) way to support tracing reconciles.